core/meterer/onchain_state.go

```diff
@@ -12,82 +12,111 @@ import (
-	GetOnDemandPaymentByAccount(ctx context.Context, blockNumber uint32, accountID string) (core.OnDemandPayment, error)
+	RefreshOnchainPaymentState(ctx context.Context, tx *eth.Transactor) error
+	GetActiveReservations(ctx context.Context) (map[string]core.ActiveReservation, error)
+	GetActiveReservationByAccount(ctx context.Context, accountID string) (core.ActiveReservation, error)
+	GetOnDemandPayments(ctx context.Context) (map[string]core.OnDemandPayment, error)
+	GetOnDemandPaymentByAccount(ctx context.Context, accountID string) (core.OnDemandPayment, error)
+	GetOnDemandQuorumNumbers(ctx context.Context) ([]uint8, error)
 }
 
 type OnchainPaymentState struct {
 	tx *eth.Transactor
 
-	ActiveReservations map[string]core.ActiveReservation
-	OnDemandPayments   map[string]core.OnDemandPayment
+	ActiveReservations    map[string]core.ActiveReservation
+	OnDemandPayments      map[string]core.OnDemandPayment
+	OnDemandQuorumNumbers []uint8
 }
 
 func NewOnchainPaymentState(ctx context.Context, tx *eth.Transactor) (OnchainPaymentState, error) {
-	activeReservations, onDemandPayments, err := CurrentOnchainPaymentState(ctx, tx)
+	blockNumber, err := tx.GetCurrentBlockNumber(ctx)
+	if err != nil {
+		return OnchainPaymentState{}, err
+	}
+
+	quorumNumbers, err := tx.GetRequiredQuorumNumbers(ctx, blockNumber)
 	if err != nil {
-		return OnchainPaymentState{tx: tx}, err
+		return OnchainPaymentState{}, err
 	}
 
 	return OnchainPaymentState{
-		tx:                 tx,
-		ActiveReservations: activeReservations,
-		OnDemandPayments:   onDemandPayments,
+		tx:                    tx,
+		ActiveReservations:    make(map[string]core.ActiveReservation),
+		OnDemandPayments:      make(map[string]core.OnDemandPayment),
+		OnDemandQuorumNumbers: quorumNumbers,
 	}, nil
 }
 
-// CurrentOnchainPaymentState returns the current onchain payment state (TODO: can optimize based on contract interface)
-func CurrentOnchainPaymentState(ctx context.Context, tx *eth.Transactor) (map[string]core.ActiveReservation, map[string]core.OnDemandPayment, error) {
+// RefreshOnchainPaymentState returns the current onchain payment state (TODO: can optimize based on contract interface)
+func (pcs *OnchainPaymentState) RefreshOnchainPaymentState(ctx context.Context, tx *eth.Transactor) error {
 	blockNumber, err := tx.GetCurrentBlockNumber(ctx)
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
 
-	activeReservations, err := tx.GetActiveReservations(ctx, blockNumber)
-	if err != nil {
-		return nil, nil, err
+	accountIDs := make([]string, 0, len(pcs.ActiveReservations))
+	for accountID := range pcs.ActiveReservations {
+		accountIDs = append(accountIDs, accountID)
 	}
 
-	onDemandPayments, err := tx.GetOnDemandPayments(ctx, blockNumber)
+	activeReservations, err := tx.GetActiveReservations(ctx, blockNumber, accountIDs)
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
+	pcs.ActiveReservations = activeReservations
```

Comment:
I think we should lock this method unless there is a strong reason not to, as i don't think it will cause any performance issues but there is a chance `ActiveReservations` and `OnDemandPayments` are accessed in parallel?